### ISSUE TITLE
ssl error during singularity pull of images from self-hosted registry

### ISSUE THREAD

**Issue by hcraT** [apptainer/singularity#849]
## Version of Singularity:

2.2.99 (git development branch)

### Expected behavior

Singularity image for requested docker image is generated.

### Actual behavior

The following error is produced:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/urllib/request.py", line 1318, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/usr/lib/python3.6/http/client.py", line 1400, in connect
    server_hostname=server_hostname)
  File "/usr/lib/python3.6/ssl.py", line 401, in wrap_socket
    _context=self, _session=session)
  File "/usr/lib/python3.6/ssl.py", line 808, in __init__
    self.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 1061, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.6/ssl.py", line 683, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:748)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/libexec/singularity/python/size.py", line 90, in <module>
    main()
  File "/usr/libexec/singularity/python/size.py", line 82, in main
    contentfile=LAYERFILE)
  File "/usr/libexec/singularity/python/docker/main.py", line 57, in SIZE
    client = DockerApiConnection(image=image,auth=auth)
  File "/usr/libexec/singularity/python/docker/api.py", line 82, in __init__
    self.load_image(kwargs['image'])
  File "/usr/libexec/singularity/python/docker/api.py", line 118, in load_image
    self.update_token()
  File "/usr/libexec/singularity/python/docker/api.py", line 129, in update_token
    response = self.get_tags(return_response=True)
  File "/usr/libexec/singularity/python/docker/api.py", line 204, in get_tags
    response = self.get(base)
  File "/usr/libexec/singularity/python/base.py", line 246, in get
    response = self.submit_request(request,return_response=return_response)
  File "/usr/libexec/singularity/python/base.py", line 263, in submit_request
    response = urlopen(request)
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/usr/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 1361, in https_open
    context=self._context, check_hostname=self._check_hostname)
  File "/usr/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:748)>
```

### Steps to reproduce behavior

Run `singularity pull docker://server:5000/image` on a a registry that doesn't support https.

### Possbile fix

I found that the problem derives from the usage of the function add_http [here](https://github.com/singularityware/singularity/blob/master/libexec/python/docker/api.py). The function add_http by default adds the https schema.
My idea would be to capture the exception and fallback to http (maybe only if the server is localhost or in specific cases that would not constitute a security risk).

**Comment by vsoch** [comment]
hey @hcraT ! I've done a PR against development to test fixing this issue: https://github.com/singularityware/singularity/pull/850 If you want to clone the branch to test it out, would be greatly appreciated!

```
git checkout -b vsoch-add/https-optional development
git pull https://github.com/vsoch/singularity.git add/https-optional
```
then install per usual

**Comment by hcraT** [comment]
Thank @vsoch tried it and it works correctly!

**Comment by mimame** [comment]
Hi @vsoch

This fix works with _pull_ and _import_ but it doesn't work for _bootstrap_ command  :(

> cat test_bootstrapping (also I tried step 3 of [https://github.com/singularityware/singularity/issues/429#issuecomment-269085751](https://github.com/singularityware/singularity/issues/429#issuecomment-269085751))
> 
> ```
> BootStrap: docker
> From localhost:5000/test/test:latest
> %post
>    echo "Wiiii"   
> ```

```
export SINGULARITY_DOCKER_REGISTRY='--registry localhost:5000'
export SINGULARITY_NOHTTPS='yes'
singularity -F -s 4096 test.img
sudo -E singularity --debug bootstrap -f test.img test_bootstrapping
```

Thanks in advance.

**Comment by vsoch** [comment]
could you please show output with `--debug`?

**Comment by mimame** [comment]
Sorry, I forgot to attach it.
[bootstrapping_debug.txt](https://github.com/singularityware/singularity/files/1205395/bootstrapping_debug.txt)

**Comment by vsoch** [comment]
hey @mimadrid it looks like it's being unset:

```
[0mDEBUG   [U=0,P=31786]      envclean()                                Unsetting environment variable: SINGULARITY_DOCKER_REGISTRY
[0mDEBUG   [U=0,P=31786]      envclean()                                Unsetting environment variable: SINGULARITY_DOCKER_REGISTRY
```

Do you want to try adding it as Registry: in the bootstrap header?

**Comment by mimame** [comment]
@vsoch Sorry but I don't know which file is :/

**Comment by vsoch** [comment]
ok, I think I know what is happening. With bootstrap (and generally) variables that are prepended with `SINGULARITY_` are chopped and then added to the registry as the remainder. So for pull and import, `SINGULARITY_NOHTTPS` is added to the registry as `NOHTTPS`. This works for pull and import because the original variables are still usable in the environment. However, with bootstrap, we clean the environment so we **only** have the `NOHTTPS` version and so the script detects that it is set to False because `NOHTTPS` isn't searched for. So the quick fix would be to have it search for `NOHTTPS`

https://github.com/vsoch/singularity/tree/fix/no-https

Would you mind testing this branch? If it works for all three, I'll do a PR. If this doesn't work, we can either check for both, or have some carrying forward of the variable (as an exception to clean).

**Comment by mimame** [comment]
I though it only happened with variables with SINGULARITYENV_ as prefix but I still have the same problem
`VERBOSE2 NOHTTPS found as False`
The env variable is not inside the container :/
Here you are the log with your branch: [bootstrapping_bad.txt](https://github.com/singularityware/singularity/files/1206253/bootstrapping_bad.txt)

My workstation is OpenSuse Tumbleweed but my personal laptop is Arch so in twelve hours I will check the branch there. 

Thanks a lot for your attention :)

**Comment by vsoch** [comment]
I think it means that even the registry isn't properly exporting it. For example, here is a quick way to test:

```
cd singularity/libexec/python
NOHTTPS=yes python -c 'import defaults'

VERBOSE2 SINGULARITY_COMMAND_ASIS found as False
VERBOSE2 SINGULARITY_ROOTFS not defined (None)
VERBOSE2 SINGULARITY_METADATA_FOLDER found as None/.singularity.d
VERBOSE2 SINGULARITY_FIX_PERMS found as False
VERBOSE2 SINGULARITY_COLORIZE not defined (None)
VERBOSE2 SINGULARITY_DISABLE_CACHE found as False
VERBOSE2 SINGULARITY_CACHEDIR found as /home/vanessa/.singularity
VERBOSE2 SINGULARITY_ENVIRONMENT found as None/.singularity.d/environment
VERBOSE2 SINGULARITY_RUNSCRIPT found as None/singularity
VERBOSE2 SINGULARITY_TESTFILE found as None/.singularity.d/test
VERBOSE2 SINGULARITY_DEFFILE found as None/.singularity.d/Singularity
VERBOSE2 SINGULARITY_HELPFILE found as None/.singularity.d/runscript.help
VERBOSE2 SINGULARITY_ENVBASE found as None/.singularity.d/env
VERBOSE2 SINGULARITY_LABELFILE found as None/.singularity.d/labels.json
VERBOSE2 SINGULARITY_INCLUDECMD found as False
VERBOSE2 NOHTTPS found as yes
VERBOSE2 SINGULARITY_PULLFOLDER found as /home/vanessa/Documents/Dropbox/Code/singularity/singularity/libexec/python
VERBOSE2 SHUB_NAMEBYHASH not defined (None)
VERBOSE2 SHUB_NAMEBYCOMMIT not defined (None)
VERBOSE2 SHUB_CONTAINERNAME not defined (None)
VERBOSE2 SINGULARITY_CONTENTS found as None/.singularity.d/.layers
VERBOSE2 SINGULARITY_PYTHREADS found as 9
```

I just added an export for `NOHTTPS` to the bootstrap script - give it one more try!

**Comment by mimame** [comment]
NOHTTPS=yes python -c 'import defaults' give me the same output than you and with the new commit I have also the same problem, Arch included :(

**Comment by vsoch** [comment]
Reported fixed by @mimadrid in #1116 closing!
